qiskit/passmanager/flow_controllers.py

```diff
@@ -133,6 +133,8 @@ def iter_tasks(self, state: PassManagerState) -> Generator[Task, PassManagerStat
                 state = yield task
             if not self.do_while(state.property_set):
                 return
+            # Remove stored tasks from the completed task collection for next loop
+            state.workflow_status.completed_passes.difference_update(self.tasks)
```

Comment:
I kind of get why we're doing this, but the need also potentially points to a data-model problem in the idea of "completed passes" to me; the pass _still_ completely after a loop iteration, and it's not immediately clear to me that starting a new loop should throw that information away.

What part of the logic is causing a pass to get skipped if it was completed?  Passes aren't required to be idempotent, so a pass having been run shouldn't be sufficient to cause it to be skipped on a second attempt.

Reply:
This line. Change to this logic causes many test failures in `test.python.transpiler.test_pass_scheduler`
https://github.com/Qiskit/qiskit/blob/dabc5e60df8e030df7f63493f00a8f5ccd5f223f/qiskit/passmanager/base_tasks.py#L95-L99

I guess this protection is added because of `GenericPass.required`. i.e. 
```
pm = PassManager([PassB, PassA[required=PassB]])
```
runs `PassB` twice.

Reply:
Ok, but I think that's potentially indicative of buggy handling: `pass_ = MyPass(); PassManager([pass_, pass_])` should run `pass_` twice unless the pass specifically marks itself as being idempotent somehow.  It's a different story if the pass is inserted into the pipeline by  `requires`, because that's an implicit add and it's fine to re-use a previous run if we know that the output from it is still valid.

Reply:
I don't really like the required mechanism. Since dependencies are needed to be instantiated within the target pass, the target pass may require extra constructor args if passes have different interface. If we remove this mechanism, I think we can drop `completed_passes` and awkward pass equality check.

Reply:
We definitely do use the `required` handling in some places in the existing pipelines, so it won't be as easy as completely stripping it out.  I think the intent of it is good, it's just that it uses referential pass equality to implicitly mean "idempotent", which isn't generally true, and that's the knock-on problem here.

Reply:
Actually circuit transform pass is not idempotent (so `pass_ = MyPass(); PassManager([pass_, pass_])` should work as you expect), but it also removes all passes because `.preserves` usually empty at least in Qiskit passes.
https://github.com/Qiskit/qiskit/blob/dabc5e60df8e030df7f63493f00a8f5ccd5f223f/qiskit/transpiler/basepasses.py#L229-L237

Reply:
Yeah, that's what I mean - I think `PassManager([pass_, pass_])` should run `pass_` twice in the general case, because there's nothing marking `_pass` as idempotent, but the "has this pass run?" handling that existed in 0.25.3 would mean it only runs once.  Since it was in 0.25.3 and the 0.45.0 passmanager just maintains the same behaviour, there's no need to rush in a fix now.